module/zfs/dbuf.c

```diff
@@ -683,6 +683,18 @@ dbuf_evict_one(void)
 	    multilist_sublist_t *, mls);
 
 	if (db != NULL) {
+		/*
+		 * Bail out if another process is already evicting
+		 * this dbuf's objset.
+		 */
+		uint64_t tid = (uint64_t)(uintptr_t)curthread;
+		objset_t *os = db->db_objset;
+		(void) atomic_cas_64(&os->os_ev_tid, 0, tid);
+		if (os->os_ev_tid != tid) {
+			multilist_sublist_unlock(mls);
+			mutex_exit(&db->db_mtx);
```

Comment:
you should drop this lock before the subist lock to ensure no deadlocks are possible.